File: Strivers-A2Z-DSA-Sheet/14. Dynamic Programming/4. DP on Subsequences/javascript/08. Target Sum.js
/*
=============================================================================
  QUESTION: 494. Target Sum (LeetCode)
=============================================================================

  Assign + or - to each element. Count ways to reach target sum.

  Example: nums=[1,1,1,1,1], target=3 → 5 ways
    -1+1+1+1+1, +1-1+1+1+1, +1+1-1+1+1, +1+1+1-1+1, +1+1+1+1-1

=============================================================================
  KEY INSIGHT: Reduce to Count Subsets with Sum
=============================================================================

  Split nums into two groups: P (positive) and N (negative).
  P - N = target
  P + N = total
  → P = (total + target) / 2

  So: count subsets with sum = (total + target) / 2.

  Edge cases:
    |target| > total → impossible
    (total + target) is odd → impossible

=============================================================================
  APPROACH: Space Optimized — O(N×S) Time, O(S) Space
=============================================================================
*/

function findTargetSumWays(nums, target) {
    let total = nums.reduce((a, b) => a + b, 0);
    if (Math.abs(target) > total) return 0;
    if ((total + target) % 2 !== 0) return 0;
    let s = (total + target) / 2;

    let n = nums.length;
    let prev = new Array(s + 1).fill(0);
    // Handle zeros: 0 can be +0 or -0, both contribute
    if (nums[0] === 0) prev[0] = 2;
    else { prev[0] = 1; if (nums[0] <= s) prev[nums[0]] = 1; }

    for (let i = 1; i < n; i++) {
        let curr = new Array(s + 1).fill(0);
        for (let t = 0; t <= s; t++) {
            let take = nums[i] <= t ? prev[t - nums[i]] : 0;
            curr[t] = take + prev[t];
        }
        prev = curr;
    }
    return prev[s];
}

// ==========================================================================
// DRIVER CODE
// ==========================================================================
console.log(findTargetSumWays([1, 1, 1, 1, 1], 3));  // 5
console.log(findTargetSumWays([1], 1));                // 1

File: Strivers-A2Z-DSA-Sheet/14. Dynamic Programming/4. DP on Subsequences/python/08. Target Sum.py
"""
  QUESTION: 494. Target Sum (LeetCode)
  Assign +/- to each element. Count ways to reach target.
  Reduces to: count subsets with sum = (total + target) / 2
  Example: nums=[1,1,1,1,1], target=3 → 5
"""

def find_target_sum_ways(nums, target):
    total = sum(nums)
    if abs(target) > total or (total + target) % 2 != 0: return 0
    s = (total + target) // 2
    n = len(nums)
    prev = [0]*(s+1)
    if nums[0] == 0: prev[0] = 2
    else:
        prev[0] = 1
        if nums[0] <= s: prev[nums[0]] = 1
    for i in range(1, n):
        curr = [0]*(s+1)
        for t in range(s+1):
            take = prev[t-nums[i]] if nums[i] <= t else 0
            curr[t] = take + prev[t]
        prev = curr
    return prev[s]

if __name__ == "__main__":
    print(find_target_sum_ways([1,1,1,1,1], 3))  # 5

